File: nodejs-typescript/src/domain/CreditPolicy.ts
import type { Money } from './Money.js';

/** Applies a credit to an account balance (parity with .NET ICreditPolicy). */
export interface CreditPolicy {
  applyCredit(currentBalance: Money, amount: Money): Money;
}

export class DefaultCreditPolicy implements CreditPolicy {
  private static readonly instance = new DefaultCreditPolicy();

  static getInstance(): DefaultCreditPolicy {
    return DefaultCreditPolicy.instance;
  }

  applyCredit(currentBalance: Money, amount: Money): Money {
    return currentBalance.add(amount);
  }
}

File: ruby/lib/mable_bank/domain/credit_policy.rb
# frozen_string_literal: true

module MableBank
  module Domain
    # Applies credits to an account balance (parity with .NET ICreditPolicy).
    module CreditPolicy
      def apply_credit(current_balance, amount)
        raise NotImplementedError, "#{self.class} must implement #apply_credit"
      end
    end

    class DefaultCreditPolicy
      include CreditPolicy

      def apply_credit(current_balance, amount)
        current_balance + amount
      end
    end
  end
end
